autoload/ale/highlight.vim

```diff
@@ -49,12 +67,83 @@ endfunction
 
 " Given a loclist for current items to highlight, remove all highlights
 " except these which have matching loclist item entries.
+
 function! ale#highlight#RemoveHighlights() abort
-    for l:match in getmatches()
-        if l:match.group =~# '^ALE'
-            call matchdelete(l:match.id)
+    if s:nvim_api
+        if get(b:, 'ale_nvim_highlight_id', 0)
+            let l:bufnr = bufnr('%')
+            " NOTE: 0, -1 means from 0 line till the end of buffer
+            call nvim_buf_clear_namespace(l:bufnr, b:ale_nvim_highlight_id, 0, -1)
         endif
-    endfor
+    else
+        for l:match in getmatches()
+            if l:match.group =~# '^ALE'
+                call matchdelete(l:match.id)
+            endif
+        endfor
+    endif
+endfunction
+
+function! s:highlight_line(bufnr, lnum, group) abort
+    if s:nvim_api
+        let l:highlight_id = s:ale_nvim_highlight_id(a:bufnr)
+        call nvim_buf_add_highlight(
+            \ a:bufnr, l:highlight_id, a:group,
+            \ a:lnum, 0, -1)
+    else
+        call matchaddpos(a:group, [a:lnum])
+    endif
+endfunction
+
+function! s:highlight_range(bufnr, range, group) abort
+    if s:nvim_api
+        let l:highlight_id = s:ale_nvim_highlight_id(a:bufnr)
+        " NOTE: lines and columns indicies are 0-based in nvim_buf_* API.
+        let l:lnum = a:range.lnum - 1
+        let l:end_lnum = a:range.end_lnum - 1
+        let l:col = a:range.col - 1
+        let l:end_col = a:range.end_col
+
+        if l:lnum >= l:end_lnum
+            " For single lines, just return the one position.
+            call nvim_buf_add_highlight(
+                \ a:bufnr, l:highlight_id, a:group,
+                \ l:lnum, l:col, l:end_col)
+        else
+            " highlight first line from start till the line end
+            call nvim_buf_add_highlight(
+                \ a:bufnr, l:highlight_id, a:group,
+                \ l:lnum, l:col, -1)
+
+            " highlight all lines between the first and last entirely
+            let l:cur = l:lnum + 1
+
+            while l:cur < l:end_lnum
+                call nvim_buf_add_highlight(
+                        \ a:bufnr, l:highlight_id, a:group,
+                        \ l:cur, 0, -1
+                        \ )
+                let l:cur += 1
+            endwhile
+
+            call nvim_buf_add_highlight(
+                \ a:bufnr, l:highlight_id, a:group,
+                \ l:end_lnum, 0, l:end_col
+                \ )
```

Comment:
Put line continuation characters on the same level of indentation as previous lines, and up-indent the content within. See most of the code in the codebase for examples.

Reply:
Fixed. We'll need a lint rule for that!